internal/metrics/resource/resources.go

```diff
@@ -0,0 +1,96 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/rhobs/multicluster-observability-addon/internal/addon"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+var (
+	mu          sync.Mutex
+	initialized bool
+	ErrNotOwner = fmt.Errorf("controller is not the owner")
+)
+
+func DeployDefaultResourcesOnce(ctx context.Context, c client.Client, logger logr.Logger, ns string) error {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if initialized {
+		return nil
+	}
+
+	logger.Info("deploying default monitoring resources")
+
+	// Get clusterManagementAddon resource to use as owner
+	owner := &addonapiv1alpha1.ClusterManagementAddOn{}
+	if err := c.Get(ctx, types.NamespacedName{Name: addon.Name, Namespace: ns}, owner); err != nil {
+		return err
+	}
+
+	// Deploy default resources
+	resources := DefaultPlaftformAgentResources(ns)
+	resources = append(resources, DefaultUserWorkloadAgentResources(ns)...)
+	for _, resource := range resources {
+		if err := CreateOrUpdateResource(ctx, c, resource, owner, logger); err != nil {
+			return fmt.Errorf("failed to create or update resource %s: %w", resource.GetName(), err)
+		}
+	}
+
+	initialized = true
+
+	return nil
+}
+
+func CreateOrUpdateResource(ctx context.Context, c client.Client, newResource, owner client.Object, logger logr.Logger) error {
```

Comment:
For this function I would refactor it to be more aligned with  https://github.com/stolostron/multicluster-observability-addon/pull/88/files#diff-d02abdba763a399da60ae895a308ea10b5ed0102c12cee1c8b497c4c2c99a527R62 since it's a simpler approach

Reply:
Nice this createOrUpdate function from the controller runtime. I did not know it existed. I refactored the code.